### ISSUE TITLE
"incorrect numbers of arguments" in codegen

### ISSUE THREAD

**Issue by gxyd** [lfortran/lfortran#3632]
```f90
module calculator_mod
    implicit none
    private
    public :: calculator

    type :: calculator
    contains
        procedure :: SQUARE
    end type calculator

contains

    function SQUARE(this, x)
        class(calculator), intent(in) :: this
        integer, optional, intent(in) :: x
        integer :: SQUARE

        if (present(x)) then
            square = x * x
        else
            square = 1
        end if
    end function SQUARE

end module calculator_mod

program CalculateSquare
    use calculator_mod
    implicit none
    type(calculator) :: calc
    integer :: number

    number = 4

    print *, 'The square of ', number, ' is ', calc%SQUARE(number)
    print *, 'The default square is ', calc%SQUARE()
end program CalculateSquare
```

compiling it with LFortran gives an error:

```bash
; ModuleID = 'LFortran'
source_filename = "LFortran"

%calculator_polymorphic = type { i64, %calculator* }
%calculator = type {}
%__vtab_calculator = type { i64 }

@0 = private unnamed_addr constant [2 x i8] c" \00", align 1
@1 = private unnamed_addr constant [2 x i8] c"\0A\00", align 1
@2 = private unnamed_addr constant [15 x i8] c"The square of \00", align 1
@3 = private unnamed_addr constant [5 x i8] c" is \00", align 1
@4 = private unnamed_addr constant [17 x i8] c"%s%s%d%s%s%s%d%s\00", align 1
@5 = private unnamed_addr constant [2 x i8] c" \00", align 1
@6 = private unnamed_addr constant [2 x i8] c"\0A\00", align 1
@7 = private unnamed_addr constant [23 x i8] c"The default square is \00", align 1
@8 = private unnamed_addr constant [9 x i8] c"%s%s%d%s\00", align 1

define i32 @__module_calculator_mod_square(%calculator_polymorphic* %this, i32* %x, i1* %is_x_present_) {
.entry:
  %0 = alloca %__vtab_calculator, align 8
  %1 = getelementptr %__vtab_calculator, %__vtab_calculator* %0, i32 0, i32 0
  store i64 0, i64* %1, align 4
  %square = alloca i32, align 4
  %2 = load i1, i1* %is_x_present_, align 1
  br i1 %2, label %then, label %else

then:                                             ; preds = %.entry
  %3 = load i32, i32* %x, align 4
  %4 = load i32, i32* %x, align 4
  %5 = mul i32 %3, %4
  store i32 %5, i32* %square, align 4
  br label %ifcont

else:                                             ; preds = %.entry
  store i32 1, i32* %square, align 4
  br label %ifcont

ifcont:                                           ; preds = %else, %then
  br label %return

return:                                           ; preds = %ifcont
  %6 = load i32, i32* %square, align 4
  ret i32 %6
}

define i32 @main(i32 %0, i8** %1) {
.entry:
  call void @_lpython_set_argv(i32 %0, i8** %1)
  %calc = alloca %calculator, align 8
  %number = alloca i32, align 4
  store i32 4, i32* %number, align 4
  %2 = load i32, i32* %number, align 4
  %3 = alloca %calculator_polymorphic, align 8
  %4 = getelementptr %calculator_polymorphic, %calculator_polymorphic* %3, i32 0, i32 0
  store i64 0, i64* %4, align 4
  %5 = getelementptr %calculator_polymorphic, %calculator_polymorphic* %3, i32 0, i32 1
  store %calculator* %calc, %calculator** %5, align 8
  %6 = call i32 @__module_calculator_mod_square(%calculator_polymorphic* %3, i32* %number)
  call void (i8*, ...) @_lfortran_printf(i8* getelementptr inbounds ([17 x i8], [17 x i8]* @4, i32 0, i32 0), i8* getelementptr inbounds ([15 x i8], [15 x i8]* @2, i32 0, i32 0), i8* getelementptr inbounds ([2 x i8], [2 x i8]* @0, i32 0, i32 0), i32 %2, i8* getelementptr inbounds ([2 x i8], [2 x i8]* @0, i32 0, i32 0), i8* getelementptr inbounds ([5 x i8], [5 x i8]* @3, i32 0, i32 0), i8* getelementptr inbounds ([2 x i8], [2 x i8]* @0, i32 0, i32 0), i32 %6, i8* getelementptr inbounds ([2 x i8], [2 x i8]* @1, i32 0, i32 0))
  %7 = alloca %calculator_polymorphic, align 8
  %8 = getelementptr %calculator_polymorphic, %calculator_polymorphic* %7, i32 0, i32 0
  store i64 0, i64* %8, align 4
  %9 = getelementptr %calculator_polymorphic, %calculator_polymorphic* %7, i32 0, i32 1
  store %calculator* %calc, %calculator** %9, align 8
  %10 = alloca i32, align 4
  %11 = call i32 @__module_calculator_mod_square(%calculator_polymorphic* %7, i32* %10)
  call void (i8*, ...) @_lfortran_printf(i8* getelementptr inbounds ([9 x i8], [9 x i8]* @8, i32 0, i32 0), i8* getelementptr inbounds ([23 x i8], [23 x i8]* @7, i32 0, i32 0), i8* getelementptr inbounds ([2 x i8], [2 x i8]* @5, i32 0, i32 0), i32 %11, i8* getelementptr inbounds ([2 x i8], [2 x i8]* @6, i32 0, i32 0))
  ret i32 0
}

declare void @_lpython_set_argv(i32, i8**)

declare void @_lfortran_printf(i8*, ...)
code generation error: asr_to_llvm: module failed verification. Error:
Incorrect number of arguments passed to called function!
  %6 = call i32 @__module_calculator_mod_square(%calculator_polymorphic* %3, i32* %number)
Incorrect number of arguments passed to called function!
  %11 = call i32 @__module_calculator_mod_square(%calculator_polymorphic* %7, i32* %10)



Note: Please report unclear or confusing messages as bugs at
https://github.com/lfortran/lfortran/issues.
```

**Comment by gxyd** [comment]
The interesting part is, the same program without making "SQUARE" to be a type-bound procedure, works just fine (i.e.):
```f90
program CalculateSquare
    implicit none
    integer :: number, result

    number = 4
    result = SQUARE(number)

    print *, 'The square of ', number, ' is ', SQUARE(number)

contains

    function SQUARE(x)
        implicit none
        integer, optional, intent(in) :: x
        integer :: SQUARE

        if (present(x)) then
            SQUARE = x * x
        else
            SQUARE = 1
        end if
    end function SQUARE

end program CalculateSquare
```

the above program works just fine with LFortran.

**Comment by gxyd** [comment]
I think we can close this issue, as it was fixed in #3711 , and it already added test case for it as well.
